blockstack_client/backend/drivers/ipfs.py

```diff
@@ -0,0 +1,410 @@
+          url = r['Path']
+        else:
+          url = chunk_path
+        compressed_data = api.cat(url)
+        end = time.time()
+        size = len(compressed_data)
+
+        try:
+            data = decompress_chunk( compressed_data )
+        except:
+            data = compressed_data
+        
+    except Exception, e:
+        log.error("Failed to read '%s'" % chunk_path)
+        log.exception(e)
+        
+    if os.environ.get("BLOCKSTACK_TEST") == "1":
+        log.debug("[BENCHMARK] s3.read_chunk %s: %s" % (size, end - begin))
+    return data
+
+#-------------------------
+def delete_chunk( chunk_path, is_mutable ):
+    """
+    Delete a chunk of data from IPFS.
+    
+    Return True on success 
+    Return False on error.
+    """
+    
+    api = ipfsapi.connect(IPFS_SERVER, IPFS_PORT)
+
+    rc = True
+
+    if is_mutable:
+      r = api.name_resolve( chunk_path )
+      url = r['Path']
+    else:
+      url = chunk_path
+
+    try:
+        api.pin_rm( url )
+    except Exception, e:
+        log.error("Failed to delete '%s'" % chunk_path)
+        log.exception(e)
+        rc = False
+
+    return rc
+
+
+def storage_init(conf, **kwargs):
+   """
+   Initialize IPFS storage driver
+   """
+   global IPFS_SERVER, IPFS_PORT, IPFS_COMPRESS
+
+   # path to the CLI's configuration file (where you can stash driver-specific configuration)
+   config_path = conf['path']
+   if os.path.exists( config_path ):
+
+       parser = SafeConfigParser()
+        
+       try:
+           parser.read(config_path)
+       except Exception, e:
+           log.exception(e)
+           return False
+
+       if parser.has_section('ipfs'):
+            
+            if parser.has_option('ipfs', 'server'):
+                IPFS_SERVER = parser.get('ipfs', 'server')
+                
+            if parser.has_option('ipfs', 'port'):
+                IPFS_PORT = parser.get('ipfs', 'port')
+            
+            if parser.has_option('ipfs', 'compress'):
+                AWS_COMPRESS = (parser.get('ipfs', 'compress', 'false').lower() in ['true', '1'])
+
+   return True 
+
+
+def handles_url( url ):
+    """
+    Does this storage driver handle this kind of URL?
+
+    It is okay if other drivers say that they can handle it.
+    This is used by the storage system to quickly filter out
+    drivers that don't handle this type of URL.
+
+    A common strategy is simply to check if the scheme
+    matches what your driver does.  Another common strategy
+    is to check if the URL matches a particular regex.
+    """
+    pattern = re.compile("Qm[1-9A-HJ-NP-Za-km-z]{44}")
+    if pattern.match(url):
+      return True
+    else:
+      return False
+
+
+def make_mutable_url( data_id ):
+    """
+    Get data by URL
+    """
+    api = ipfsapi.connect(IPFS_SERVER, IPFS_PORT)
+
+    key = 'mutable-{}'.format(data_id.replace( "/", r"\x2f" ))
+
+    ipfs_key_gen( key )
+   
+    r = api.name_publish( hash_data(''), opts = {'key':key})
```

Comment:
Instead of publishing many times for each write, you could use the files api (`ipfs files`) to put everything inside a virtual mutable tree, and then periodically do an ipns publish with the root hash of that.